File: src/utils/storage/DefDataStorage.ts
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { DataStorage } from "./declare/DataStorage";

export function DefDataStorage<SchemaStorage extends object>(defaultValues: SchemaStorage, storageName = 'storage'): DataStorage<SchemaStorage> {
    type Keys = keyof typeof defaultValues;
    const cache = { ...defaultValues };
    const prefix = storageName + '_';
    load();

    function set<T extends Keys>(key: T, value: SchemaStorage[T]) {
        cache[key] = value;
        Storage.set(prefix + (key as string), json.encode({ value: value }, {}));
    }

    function get<T extends Keys>(key: T): SchemaStorage[T] {
        return cache[key];
    }

    function load() {
        Object.keys(defaultValues).forEach(key => _loadItem(key));
    }

    function _loadItem(key: string) {
        let storageValue = Storage.get(prefix + key);
        if (storageValue == undefined) {
            storageValue =json.encode({ value: defaultValues[key as Keys] }, {});
            Storage.set(prefix + key, storageValue);
        }
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-argument
        cache[key as Keys] = json.decode(storageValue, {}).value;
    }

    function clear() {
        Object.keys(cache).forEach(key => {
            set(key as Keys, defaultValues[key as Keys]);
        });
    }

    return {
        set, get, load, clear, schemaObject: defaultValues
    };
}

File: app/utils/storage/DefDataStorage.lua
local ____lualib = require("lualib_bundle")
local __TS__ObjectAssign = ____lualib.__TS__ObjectAssign
local __TS__ObjectKeys = ____lualib.__TS__ObjectKeys
local __TS__ArrayForEach = ____lualib.__TS__ArrayForEach
local ____exports = {}
function ____exports.DefDataStorage(defaultValues, storageName)
    if storageName == nil then
        storageName = "storage"
    end
    local load, _loadItem, cache, prefix
    function load()
        __TS__ArrayForEach(
            __TS__ObjectKeys(defaultValues),
            function(____, key) return _loadItem(key) end
        )
    end
    function _loadItem(key)
        local storageValue = Storage.get(prefix .. key)
        if storageValue == nil then
            storageValue = json.encode({value = defaultValues[key]}, {})
            Storage.set(prefix .. key, storageValue)
        end
        cache[key] = json.decode(storageValue, {}).value
    end
    cache = __TS__ObjectAssign({}, defaultValues)
    prefix = storageName .. "_"
    load()
    local function set(key, value)
        cache[key] = value
        Storage.set(
            prefix .. key,
            json.encode({value = value}, {})
        )
    end
    local function get(key)
        return cache[key]
    end
    local function clear()
        __TS__ArrayForEach(
            __TS__ObjectKeys(cache),
            function(____, key)
                set(key, defaultValues[key])
            end
        )
    end
    return {
        set = set,
        get = get,
        load = load,
        clear = clear,
        schemaObject = defaultValues
    }
end
return ____exports
